### PR TITLE
Improve media stream handling during reconnections

### PR DESCRIPTION
**Pull request by richnew10**
This PR addresses three related bugs that can be observed when testing with a broken network:

* Repeated re-removal of transform devices wasn't safe from this calling context. One commit here makes this safe.
* `actionDisconnect` with `reconnecting = true` terminates media streams. Another commit prevents this.
* The eventual timeout of timeouts hits `[WARN] SDK - no transition found from NotConnected with Fail`, and so the final disconnect does nothing. This PR avoids entering this state entirely.

Future work is likely to shift the responsibility for device cleanup entirely to `stop`, which is vaguely the notion suggested in the docs. That change will need to be communicated clearly, because it resolves an ambiguity — today the SDK closes streams when disconnected, which is not documented anywhere and is probably undesirable.

**Testing**

> 1. Have you successfully run `npm run build:release` locally?

Only running tests by hand.

> 2. How did you test these changes?

By simulating connection establishment failures.

> 3. Can these changes be tested using the demo application? If yes, which demo application can be used to test it?

Yes; meetingV2.

> 4. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?

No.

> 5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?

No.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/audiovideocontroller/DefaultAudioVideoController.ts [hunk 1]
```diff
@@ -549,17 +551,29 @@ export default class DefaultAudioVideoController
...
+      this.logger.info(`Start failed: ${status} due to error ${error}.`);
+
+      // I am not able to successfully reach this state in the test suite with mock
+      // websockets -- it always ends up in 'Disconnecting' instead. As such, this
+      // has to be marked for Istanbul.
+      /* istanbul ignore if */
+      if (this.sessionStateController.state() === SessionStateControllerState.NotConnected) {
+        // There's no point trying to 'disconnect', because we're not connected.
+        // The session state controller will bail.
+        this.logger.info('Start failed and not connected. Not cleaning up.');
```
Comment (devalevenkatesh) [inline src/audiovideocontroller/DefaultAudioVideoController.ts:566]: Before this change, the [`notifyStop()`](https://github.com/aws/amazon-chime-sdk-js/blob/8de51639a8d00519d17d42f1722dd7fea2d065f2/src/audiovideocontroller/DefaultAudioVideoController.ts#L817) would send either of 'meetingStartFailed', 'meetingFailed', or 'meetingEnded' meeting event based on the meeting status or the meeting duration. With this change, we are returning early here and the meeting start has failed, should we publish a 'meetingStartFailed' event here?

Comment (richnew10) [reply src/audiovideocontroller/DefaultAudioVideoController.ts:566]: No it wouldn't — `perform` returns without running its argument if there is no state transition for the current state. There is no transition from `NotConnected` to `Fail`, so lines 553–560 would not run if we're in state `NotConnected`, and thus in the current codebase no event would be published.

Comment (richnew10) [reply src/audiovideocontroller/DefaultAudioVideoController.ts:566]: `[WARN] SDK - no transition found from NotConnected with Fail`, as mentioned in the PR text, indicates that the code that emits the event would not run.

Comment (devalevenkatesh) [reply src/audiovideocontroller/DefaultAudioVideoController.ts:566]: Yup right, got it, so that state even does not reach, thanks for the explanation.

Comment (devalevenkatesh) [reply src/audiovideocontroller/DefaultAudioVideoController.ts:566]: The log says start failed so thought if we should publish an event for the same. We will come to this branch of code if we fail once we have already ran the `audioVideo.start()` right?

Comment (richnew10) [reply src/audiovideocontroller/DefaultAudioVideoController.ts:566]: A failure event is emitted somewhere else:

```
ConsoleLogger.js:76 2021-05-07T17:25:42.737Z [INFO] SDK - handling status: TaskFailed
10:25:42.737 ConsoleLogger.js:70 2021-05-07T17:25:42.737Z [WARN] SDK - no transition found from NotConnected with Fail
10:25:42.738 ConsoleLogger.js:67 2021-05-07T17:25:42.738Z [ERROR] SDK - failed with status code TaskFailed and error: serial group task AudioVideoStart/d6e56e8e-4901-41d6-b020-ef57a95aa535/d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f was canceled due to subtask AudioVideoStart/d6e56e8e-4901-41d6-b020-ef57a95aa535/d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f/Timeout15000ms error: serial group task AudioVideoStart/d6e56e8e-4901-41d6-b020-ef57a95aa535/d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f/Timeout15000ms/Peer was canceled due to subtask AudioVideoStart/d6e56e8e-4901-41d6-b020-ef57a95aa535/d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f/Timeout15000ms/Peer/SubscribeAndReceiveSubscribeAckTask (once) error: serial group task Signaling was canceled due to subtask Signaling/Timeout15000ms (once) error: OpenSignalingConnectionTask got canceled while waiting to open signaling connection
10:25:42.742 ConsoleLogger.js:76 2021-05-07T17:25:42.742Z [INFO] SDK - Not canceling Signaling: state is FINISHED
10:25:42.742 ConsoleLogger.js:76 2021-05-07T17:25:42.742Z [INFO] SDK - canceling task CreatePeerConnectionTask
10:25:42.742 ConsoleLogger.js:76 2021-05-07T17:25:42.742Z [INFO] SDK - Not canceling ReceiveAudioInputTask: state is FINISHED
10:25:42.742 ConsoleLogger.js:76 2021-05-07T17:25:42.742Z [INFO] SDK - canceling task AttachMediaInputTask
10:25:42.742 ConsoleLogger.js:76 2021-05-07T17:25:42.742Z [INFO] SDK - canceling task CreateSDPTask
10:25:42.742 ConsoleLogger.js:76 2021-05-07T17:25:42.742Z [INFO] SDK - canceling task SetLocalDescriptionTask
10:25:42.742 ConsoleLogger.js:76 2021-05-07T17:25:42.742Z [INFO] SDK - canceling task FinishGatheringICECandidatesTask
10:25:42.742 ConsoleLogger.js:76 2021-05-07T17:25:42.742Z [INFO] SDK - canceling task AudioVideoStart/d6e56e8e-4901-41d6-b020-ef57a95aa535/d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f/Timeout15000ms/Peer/SubscribeAndReceiveSubscribeAckTask
10:25:42.743 meetingV2.ts:1960 [DEMO] session stopped from {"_statusCode":19}
10:25:42.743 meetingV2.ts:1960 [DEMO] resetting stats in WebRTCStatsCollector
10:25:42.743 meetingV2.ts:1960 [DEMO] Received an event: {"name":"meetingStartFailed","attributes":{"attendeeId":"d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f","browserMajorVersion":"90","browserName":"Chrome","browserVersion":"90.0.4430.72","deviceName":"Unavailable","externalMeetingId":"aaabbcdedp","externalUserId":"8a871d92#n","meetingHistory":[{"name":"audioInputSelected","timestampMs":1620408212878},{"name":"videoInputSelected","timestampMs":1620408214141},{"name":"meetingStartRequested","timestampMs":1620408220731},{"name":"videoInputUnselected","timestampMs":1620408220736},{"name":"meetingStartFailed","timestampMs":1620408342739}],"meetingId":"d6e56e8e-4901-41d6-b020-ef57a95aa535","osName":"Mac OS","osVersion":"10.15.7","sdkVersion":"2.8.0","sdkName":"amazon-chime-sdk-js","timestampMs":1620408342739,"maxVideoTileCount":0,"meetingStatus":"TaskFailed","signalingOpenDurationMs":14997,"iceGatheringDurationMs":null,"poorConnectionCount":0,"retryCount":6,"meetingErrorMessage":"serial group task AudioVideoStart/d6e56e8e-4901-41d6-b020-ef57a95aa535/d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f was canceled due to subtask AudioVideoStart/d6e56e8e-4901-41d6-b020-ef57a95aa535/d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f/Timeout15000ms error: serial group task AudioVideoStart/d6e56e8e-4901-41d6-b020-ef57a95aa535/d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f/Timeout15000ms/Peer was canceled due to subtask AudioVideoStart/d6e56e8e-4901-41d6-b020-ef57a95aa535/d4e2e44f-f90d-cf7b-7439-5c5f77a26d0f/Timeout15000ms/Peer/SubscribeAndReceiveSubscribeAckTask (once) error: serial group task Signaling was canceled due to subtask Signaling/Timeout15000ms (once) error: OpenSignalingConnectionTask got canceled while waiting to open signaling connection"}}
```

**Review by devalevenkatesh — APPROVED** [review]

**Review by ltrung — APPROVED** [review]

**Pull request merged**